nifi-nar-bundles/nifi-standard-services/nifi-dbcp-service-bundle/nifi-dbcp-service/src/test/java/org/apache/nifi/dbcp/HikariCPConnectionPoolTest.java

```diff
@@ -0,0 +1,273 @@
+        runner.setProperty(service, HikariCPConnectionPool.MIN_IDLE, "4");
+        runner.setProperty(service, HikariCPConnectionPool.MAX_CONN_LIFETIME, "1 secs");
+
+        runner.enableControllerService(service);
+
+        Assert.assertEquals(4, service.getDataSource().getMinimumIdle());
+        Assert.assertEquals(1000, service.getDataSource().getMaxLifetime());
+
+        service.getDataSource().close();
+    }
+
+    /**
+     * Test database connection using Derby. Connect, create table, insert, select, drop table.
+     */
+    @Test
+    public void testCreateInsertSelect() throws InitializationException, SQLException {
+        final TestRunner runner = TestRunners.newTestRunner(TestProcessor.class);
+        final HikariCPConnectionPool service = new HikariCPConnectionPool();
+        runner.addControllerService("test-good1", service);
+
+        // set embedded Derby database connection url
+        runner.setProperty(service, HikariCPConnectionPool.DATABASE_URL, "jdbc:derby:" + DB_LOCATION + ";create=true");
+        runner.setProperty(service, HikariCPConnectionPool.DB_USER, "tester");
+        runner.setProperty(service, HikariCPConnectionPool.DB_PASSWORD, "testerp");
+        runner.setProperty(service, HikariCPConnectionPool.DB_DRIVERNAME, "org.apache.derby.jdbc.EmbeddedDriver");
+
+        runner.enableControllerService(service);
+
+        runner.assertValid(service);
+        final DBCPService dbcpService = (DBCPService) runner.getProcessContext().getControllerServiceLookup().getControllerService("test-good1");
+        Assert.assertNotNull(dbcpService);
+        final Connection connection = dbcpService.getConnection();
+        Assert.assertNotNull(connection);
+
+        createInsertSelectDrop(connection);
+
+        connection.close(); // return to pool
+    }
+
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
+
+    /**
+     * Test get database connection using Derby. Get many times, after a while pool should not contain any available connection and getConnection should fail.
+     */
+    @Test
+    public void testExhaustPool() throws InitializationException {
+        final TestRunner runner = TestRunners.newTestRunner(TestProcessor.class);
+        final HikariCPConnectionPool service = new HikariCPConnectionPool();
+        runner.addControllerService("test-exhaust", service);
+
+        // set embedded Derby database connection url
+        runner.setProperty(service, HikariCPConnectionPool.DATABASE_URL, "jdbc:derby:" + DB_LOCATION + ";create=true");
+        runner.setProperty(service, HikariCPConnectionPool.DB_USER, "tester");
+        runner.setProperty(service, HikariCPConnectionPool.DB_DRIVERNAME, "org.apache.derby.jdbc.EmbeddedDriver");
+
+        runner.enableControllerService(service);
+
+        runner.assertValid(service);
+        final DBCPService dbcpService = (DBCPService) runner.getProcessContext().getControllerServiceLookup().getControllerService("test-exhaust");
```

Comment:
Recommend setting a static `SERVICE_ID` and reusing across all test methods for easier maintenance and avoiding errors. On further review, is it necessary to retrieve the same object as opposed to just calling the already declared `service`?
```suggestion
        final DBCPService dbcpService = (DBCPService) runner.getProcessContext().getControllerServiceLookup().getControllerService(SERVICE_ID);
```

Reply:
Sometimes I think there are slight differences so I just went with copy-paste. I can move those into constants